geps/gep-1742.md

```diff
@@ -298,16 +298,182 @@ Could not find any HTTP specific timeouts. PRs welcomed. 😊
+    response to a client HTTP request. Whether the gateway starts the timeout before
+    or after the entire client request stream has been received, is implementation dependent.
+    This field is required `Core` support.
+
+1. `timeouts.backendRequest` is a timeout for a single request from the gateway to a backend.
+    This field is optional `Extended` support. Typically used in conjuction with retry configuration,
+    if supported by an implementation.
+    Note that retry configuration will be the subject of a separate GEP (GEP-1731).
+
+```mermaid
+sequenceDiagram
+    participant C as Client
+    participant P as Proxy
+    participant U as Upstream
+    C->>P: Connection Started
+    note left of P: timeouts.request start time (min)
+    C->>P: Starts sending Request
+    C->>P: Finishes Headers
+    C->>P: Finishes request
+    note left of P: timeouts.request start time (max)
+    P->>U: Connection Started
+    note right of P: timeouts.backendRequest start time
+    P->>U: Starts sending Request
+    P->>U: Finishes request
+    P->>U: Finishes Headers
+    U->>P: Starts Response
+    U->>P: Finishes Headers
+    note right of P: timeouts.backendRequest end time
+    note left of P: timeouts.request end time
+    U->>P: Finishes Response
+    note right of P: Repeat if retry
+    P->>C: Starts Response
+    P->>C: Finishes Headers
+    P->>C: Finishes Response
+    Note right of P: Repeat if connection sharing
+    U->>C: Connection ended
+```
+
+Both timeout fields are string duration values as specified by
+[Golang time.ParseDuration](https://pkg.go.dev/time#ParseDuration) and MUST be >= 1ms
+or 0 to disable (no timeout).
+
+### GO
+
+```go
+type HTTPRouteRule struct {
+	// Timeouts defines the timeouts that can be configured for an HTTP request.
+	//
+	// Support: Core
+	//
+	// +optional
+	// <gateway:experimental>
+	Timeouts *HTTPRouteTimeouts `json:"timeouts,omitempty"`
+
+	// ...
+}
+
+// HTTPRouteTimeouts defines timeouts that can be configured for an HTTPRoute.
+// Timeout values are formatted like 1h/1m/1s/1ms as parsed by Golang time.ParseDuration
+// and MUST BE >= 1ms or 0 to disable (no timeout).
+type HTTPRouteTimeouts struct {
+	// Request specifies the duration for processing an HTTP client request after which the
+	// gateway will time out if unable to send a response.
+	//
+	// For example, setting the `rules.timeouts.request` field to the value `10s` in an
+	// `HTTPRoute` will cause a timeout if a client request is taking longer than 10 seconds
+	// to complete.
+	//
+	// This timeout is intended to cover as close to the whole request-response transaction
+	// as possible although an implementation MAY choose to start the timeout after the entire
+	// request stream has been received instead of immediately after the transaction is
+	// initiated by the client.
+	//
+	// When this field is unspecified, request timeout behavior is implementation-dependent.
+	//
+	// Support: Core
```

Comment:
NGINX doesn't support request timeouts. However, not sure how many other implementations are in the same situation. 

Is it possible to have this feature as Extended?

Reply:
I thought NGINX's `proxy_read_timeout` aligned with `Request` timeout, is it a `BackendRequest` timeout, instead? If so, maybe `BackendRequest` should be Core, and `Request` should be Extended? Is there an NGINX timeout that takes retries into account?

Reply:
> I thought NGINX's proxy_read_timeout aligned with Request timeout, is it a BackendRequest timeout, instead?

unfortunately `proxy_read_timeout` applies not to the duration of a request, but to successive read operations from the proxied server :(